content/blog/2019-09-25-making-function-calls-work.md

```diff
@@ -0,0 +1,299 @@
+
+    # Fails!
+    main {
+      v0: int = call func;
+      print v0;
+    }
+
+    func {
+      v0: int = const 100;
+      ret v0;
+    }
+
+The opposite is allowed: functions that have a declared return type *may* be called as standalone instructions. The following program succeeds and prints 100.
+
+    # Print 100
+    main {
+      call func;
+    }
+
+    func: int {
+      v0: int = const 100;
+      print v0;
+      ret v0;
+    }
+
+
+#### Optional type annotations for function parameters
+Function definitions are now more complex, and with this added syntactic complexity comes a loss of readability. To fix this, we introduce the notion of type annotations for function parameters. Since we do not do support static type checking, these type annotations serve no operational purpose. As a result, all we have to do is update the language grammar. Below is an updated version of the above program with type annotations.
+
+    # Print fib(10)
+    main {
+        v0: int = const 10;
+        fib10: int = call fib v0;
+        print fib10;
+    }
+
+    # Return true if n <= 1, false otherwise.
+    lte_one (n: int): bool {
+        one: int = const 1;
+        lto: bool = le n one;
+        ret lto;
+    }
+
+    # Return fib(10). Requires: n >= 0.
+    fib (n: int): int {
+        base: bool = call lte_one n;
+        br base return continue;
+    return:
+        ret n;
+    continue:
+        one: int = const 1;
+        prev: int = sub n one;
+        prev2: int = sub prev one;
+        fib1: int = call fib prev;
+        fib2: int = call fib prev2;
+        ans: int = add fib1 fib2;
+        ret ans;
+    }
+
+#### Nested function definitions
+In the above example, `lte_one` is just a helper function for `fib`and is not used anywhere else. To avoid cluttering the global function definitions, it would be nice to only define functions where they are useful.
+
+To do this, we introduced support for nested function definitions (i.e., function definitions within function definitions). First we add a new rule to the grammar. We add a new `instr` rule of the format `"def" func`, where `func` is the rule for normal function definitions. The `"def"` is there to avoid issues with labels (when the parser encounters `x:` it will not know if `x` is a label or a function with a return type). In the interpreter, we introduce the notion of a local function map. Previously, we had a global function map for the entire program. The local function map restricts access to a nested function to the immediate parent function.
+
+We also change the `call` operation handler to first search the local map before searching the global map. Note that this means functions can be shadowed. Below is the above program updated with nested function definitions.
```

Comment:
There's an elephant in the room here: are these closures? That is, can a nested function declaration access variables declared in its parent?

Or is it just a way to namespace the function declarations, and has nothing to do with nested functions as they appear in user-facing languages?